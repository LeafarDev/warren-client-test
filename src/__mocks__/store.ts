export default {
  userTotalInboundYield: {
    userTotalInboundYield: {
      total: 0
    }
  },
  transactions: {
    transactions: [],
    currentCreatingTransactionType: 'deposit',
    lastCreatedTransaction: null
  },
  auth: {
    token: ''
  },
  user: {
    user: {
      'id': '1aafa847-b0b4-42c9-929f-61cc75b975a5',
      'name': 'Charles Marciano',
      'email': 'charles.marciano45@gmail.com',
      'accountId': '1539c5cc-8978-41c2-9339-1e34778f5b7d',
      'createdAt': '2021-06-02T01:42:01.987Z',
      'updatedAt': null,
      'account': {
        'id': '1539c5cc-8978-41c2-9339-1e34778f5b7d',
        'accountNumber': '12345678',
        'balance': 0.0056,
        'createdAt': '2021-06-02T01:42:01.766Z',
        'updatedAt': '2021-06-02T21:42:02.128Z'
      }
    }
  },
  isLoading: false
};
