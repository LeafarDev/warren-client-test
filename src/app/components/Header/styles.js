import { Button, Toolbar } from '@material-ui/core';
import styled from 'styled-components';

export const CustomToolbar = styled(Toolbar)`
  && {
    background-color: #6a88ff;
  }
`;

export const HeaderDiv = styled.div`
  && {
    min-height: 64px;
  }
`;

export const LogoutButton = styled(Button)`
  && {
    background-color: #e61b5d;
    color: white;
  }
`;
