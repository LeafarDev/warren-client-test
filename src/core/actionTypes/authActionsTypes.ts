export const LOGIN = 'authActionTypes/LOGIN';

export interface LoginAction {
  type: typeof LOGIN;
  email: string;
  password: string;
}

export const LOGOUT = 'authActionTypes/LOGOUT';

export interface LogoutAction {
  type: typeof LOGOUT
}

export const RESET = 'authActionTypes/RESET';

export interface ResetAction {
  type: typeof RESET
}

export const LOGIN_REQUEST = 'authActionTypes/LOGIN_REQUEST';

export interface LoginRequestAction {
  type: typeof LOGIN_REQUEST;
}

export const LOGIN_SUCCESS = 'authActionTypes/LOGIN_SUCCESS';

export interface LoginSuccessAction {
  type: typeof LOGIN_SUCCESS;
  token: string;
}

export const LOGIN_FAILURE = 'authActionTypes/LOGIN_FAILURE';

export interface LoginFailureAction {
  type: typeof LOGIN_FAILURE;
  error: Error | string;
}

export type AuthAction =
  | ResetAction
  | LogoutAction
  | LoginRequestAction
  | LoginFailureAction
  | LoginAction
  | LoginSuccessAction;
