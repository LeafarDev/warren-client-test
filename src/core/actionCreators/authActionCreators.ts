import * as actions from '../actionTypes/authActionsTypes';

export function login(
  email: string,
  password: string
): actions.LoginAction {
  return {
    type: actions.LOGIN,
    email,
    password
  };
}

export function logout(): actions.LogoutAction {
  return {
    type: actions.LOGOUT
  };
}

export function reset(): actions.ResetAction {
  return {
    type: actions.RESET
  };
}

export function loginRequest(): actions.LoginRequestAction {
  return {
    type: actions.LOGIN_REQUEST
  };
}

export function loginSuccess(
  token: string
): actions.LoginSuccessAction {
  return {
    type: actions.LOGIN_SUCCESS,
    token
  };
}

export function loginFailure(
  error: Error | string
): actions.LoginFailureAction {
  return {
    type: actions.LOGIN_FAILURE,
    error
  };
}
