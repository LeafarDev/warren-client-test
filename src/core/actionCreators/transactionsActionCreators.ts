import * as actions from '../actionTypes/transactionsActionTypes';
import { ITransaction } from '../reducers/transactionsReducer';

export function getTransactions(): actions.GetTransactionsAction {
  return {
    type: actions.GET_TRANSACTIONS
  };
}

export function saveTransaction(transaction: ITransaction): actions.SaveTransactionAction {
  return {
    type: actions.SAVE_TRANSACTION,
    transaction
  };
}

export function getTransactionsRequest(): actions.GetTransactionsRequestAction {
  return {
    type: actions.GET_TRANSACTIONS_REQUEST
  };
}

export function saveTransactionRequest(): actions.SaveTransactionRequestAction {
  return {
    type: actions.SAVE_TRANSACTION_REQUEST
  };
}

export function getTransactionsSuccess(
  transactions: ITransaction[]
): actions.GetTransactionsSuccessAction {
  return {
    type: actions.GET_TRANSACTIONS_SUCCESS,
    transactions
  };
}

export function saveTransactionSuccess(
  transaction: ITransaction | null
): actions.SaveTransactionSuccessAction {
  return {
    type: actions.SAVE_TRANSACTION_SUCCESS,
    transaction
  };
}

export function currentCreatingTransaction(
  currentCreatingTransactionType: 'withdraw' | 'deposit' | 'payment'
): actions.CurrentCreatingTransactionTypeAction {
  return {
    type: actions.CURRENT_CREATING_TRANSACTION_TYPE,
    currentCreatingTransactionType
  };
}

export function getTransactionsFailure(
  error: Error | string
): actions.GetTransactionsFailureAction {
  return {
    type: actions.GET_TRANSACTIONS_FAILURE,
    error
  };
}

export function saveTransactionFailure(
  error: Error | string
): actions.SaveTransactionFailureAction {
  return {
    type: actions.SAVE_TRANSACTION_FAILURE,
    error
  };
}
