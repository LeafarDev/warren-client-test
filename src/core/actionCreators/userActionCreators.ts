import * as actions from '../actionTypes/userActionTypes';
import { IUser } from '../reducers/userReducer';

export function setUser(user: IUser): actions.SetUserAction {
  return {
    type: actions.SET_USER,
    user
  };
}

export function getUser(): actions.GetUserAction {
  return {
    type: actions.GET_USER
  };
}

export function getUserRequest(): actions.GetUserRequestAction {
  return {
    type: actions.GET_USER_REQUEST
  };
}

export function getUserSuccess(
  user: IUser
): actions.GetUserSuccessAction {
  return {
    type: actions.GET_USER_SUCCESS,
    user
  };
}

export function getUserFailure(
  error: Error | string
): actions.GetUserFailureAction {
  return {
    type: actions.GET_USER_FAILURE,
    error
  };
}
