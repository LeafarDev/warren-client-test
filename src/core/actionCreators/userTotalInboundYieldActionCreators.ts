import * as actions from '../actionTypes/userTotalInboundYieldActionTypes';
import { IUserTotalInboundYield } from '../reducers/userTotalInboundYieldReducer';

export function getUserTotalInboundYield(): actions.GetUserTotalInboundYieldTypesAction {
  return {
    type: actions.GET_USER_TOTAL_INBOUND_YIELD
  };
}

export function getUserTotalInboundYieldRequest(): actions.GetUserTotalInboundYieldTypesRequestAction {
  return {
    type: actions.GET_USER_TOTAL_INBOUND_YIELD_REQUEST
  };
}

export function getUserTotalInboundYieldSuccess(
  userTotalInboundYield: IUserTotalInboundYield
): actions.GetUserTotalInboundYieldTypesSuccessAction {
  return {
    type: actions.GET_USER_TOTAL_INBOUND_YIELD_SUCCESS,
    userTotalInboundYield
  };
}

export function getUserTotalInboundYieldFailure(
  error: Error | string
): actions.GetUserTotalInboundYieldTypesFailureAction {
  return {
    type: actions.GET_USER_TOTAL_INBOUND_YIELD_FAILURE,
    error
  };
}
