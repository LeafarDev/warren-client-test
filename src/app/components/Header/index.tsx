import * as React from 'react';
import { AppBar, Typography } from '@material-ui/core';
import { CustomToolbar, HeaderDiv, LogoutButton } from './styles';
import { useDispatch } from 'react-redux';
import { logout } from '../../../core/actionCreators/authActionCreators';

export function Header(): JSX.Element {
  const dispatch = useDispatch();

  const onLogout = () => {
    dispatch(logout());
  };

  return (
    <HeaderDiv>
      <AppBar>
        <CustomToolbar>
          <Typography variant={'h6'} color="inherit" style={{ flex: 1 }}>
            Bank of Mars
          </Typography>
          <div>
            <LogoutButton name={'logout'} variant="contained" onClick={() => onLogout()}>
              Logout
            </LogoutButton>
          </div>
        </CustomToolbar>
      </AppBar>
    </HeaderDiv>
  );
}

export default Header;
