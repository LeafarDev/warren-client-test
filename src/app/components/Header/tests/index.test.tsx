import React from 'react';
import Header from '../index';
import { fireEvent, render, screen } from '@testing-library/react';
import * as redux from 'react-redux';
import store from '../../../../__mocks__/store';
import * as authActionCreators from '../../../../core/actionCreators/authActionCreators';
import * as actions from '../../../../core/actionTypes/authActionsTypes';

jest.mock('react-redux');

describe('<Header />', () => {

  let dispatch;
  beforeEach(() => {
    dispatch = jest.spyOn(redux, 'useDispatch')
      .mockImplementation(jest.fn());
    dispatch.mockReturnValue(jest.fn());
    jest.spyOn(redux, 'useSelector')
      .mockImplementation((cb) => cb(store));
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild }
    } = render(<Header/>);
    expect(firstChild)
      .toMatchSnapshot();
  });

  it('Logout button should dispatch the logout action', async () => {
    render(<Header/>);
    const logoutButton = await screen.getAllByRole('button', {
      name: /logout/i
    });

    jest
      .spyOn(authActionCreators, 'logout')
      .mockImplementation((): actions.LogoutAction => {
        return {
          type: actions.LOGOUT
        };
      });

    fireEvent.click(logoutButton[0]);
    expect(authActionCreators.logout)
      .toHaveBeenCalledTimes(1);
  });
});
