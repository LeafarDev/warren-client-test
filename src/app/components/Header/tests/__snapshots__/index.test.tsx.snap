// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render and match the snapshot 1`] = `
<div
  class="sc-gtsrHT LNvXb"
>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary mui-fixed MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular sc-bdnxRM cHSwRG MuiToolbar-gutters"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 MuiTypography-colorInherit"
        style="flex: 1;"
      >
        Bank of Mars
      </h6>
      <div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained sc-dlnjwi kckUMR"
          name="logout"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Logout
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
</div>
`;
